sklearn/model_selection/tests/test_split.py

```diff
@@ -0,0 +1,717 @@
+    train, test = next(splits)
+    assert_array_equal(test, [0, 2])
+    assert_array_equal(train, [1, 3])
+
+    train, test = next(splits)
+    assert_array_equal(test, [1, 3])
+    assert_array_equal(train, [0, 2])
+
+    splits = StratifiedKFold(2).split(np.ones((7, 2)), [1, 1, 1, 0, 0, 0, 0])
+    train, test = next(splits)
+    assert_array_equal(test, [0, 1, 3, 4])
+    assert_array_equal(train, [2, 5, 6])
+
+    train, test = next(splits)
+    assert_array_equal(test, [2, 5, 6])
+    assert_array_equal(train, [0, 1, 3, 4])
+
+
+def test_stratified_kfold_ratios():
+    # Check that stratified kfold preserves label ratios in individual splits
+    # Repeat with shuffling turned off and on
+    n_samples = 1000
+    X = np.ones(n_samples)
+    labels = np.array([4] * int(0.10 * n_samples) +
+                      [0] * int(0.89 * n_samples) +
+                      [1] * int(0.01 * n_samples))
+
+    for cv in [StratifiedKFold(5, shuffle=s) for s in (True, False)]:
+        for train, test in cv.split(X, y=labels):
+            assert_almost_equal(np.sum(labels[train] == 4) / len(train), 0.10,
+                                2)
+            assert_almost_equal(np.sum(labels[train] == 0) / len(train), 0.89,
+                                2)
+            assert_almost_equal(np.sum(labels[train] == 1) / len(train), 0.01,
+                                2)
+            assert_almost_equal(np.sum(labels[test] == 4) / len(test), 0.10, 2)
+            assert_almost_equal(np.sum(labels[test] == 0) / len(test), 0.89, 2)
+            assert_almost_equal(np.sum(labels[test] == 1) / len(test), 0.01, 2)
+
+
+def test_kfold_balance():
+    # Check that KFold returns folds with balanced sizes
+    for i in range(11, 17):
+        kf = KFold(5).split(X=np.ones(i))
+        sizes = []
+        for _, test in kf:
+            sizes.append(len(test))
+
+        assert_true((np.max(sizes) - np.min(sizes)) <= 1)
+        assert_equal(np.sum(sizes), i)
+
+
+def test_stratifiedkfold_balance():
+    # Check that KFold returns folds with balanced sizes (only when
+    # stratification is possible)
+    # Repeat with shuffling turned off and on
+    labels = [0] * 3 + [1] * 14
+    X = np.ones(17)
+
+    for cv in [StratifiedKFold(3, shuffle=s) for s in (True, False)]:
+        for i in range(11, 17):
+            skf = cv.split(X[:i], y=labels[:i])
+            sizes = []
+            for _, test in skf:
+                sizes.append(len(test))
+
+            assert_true((np.max(sizes) - np.min(sizes)) <= 1)
+            assert_equal(np.sum(sizes), i)
+
+
+def test_shuffle_kfold():
+    # Check the indices are shuffled properly, and that all indices are
+    # returned in the different test folds
+    kf = KFold(3, shuffle=True, random_state=0)
+    ind = np.arange(300)
+
+    all_folds = None
+    for train, test in kf.split(X=np.ones(300)):
+        sorted_array = np.arange(100)
+        assert_true(np.any(sorted_array != ind[train]))
```

Comment:
It is compared against the three sets so that it makes sense for all three iterations.
It would still pass if we just shuffle inside each fold, right?
Also ``ind[train]`` is just the same as ``train``, right, and we don't need ``ind``?

Maybe we should use ``assert_false(len(intersect_1d(sorted_array, train)), len(train))``?